### PR TITLE
Remove recently finished flows cache

### PR DESCRIPTION
**Pull request by jamiesjc**
In order to scale out web server, we need to move the existing cache from web server to database so that web server can be stateless. Currently, recentlyFinishedFlows cache keeps the flows which finished within 10 mins. This info is shown on Executing Flows - Recently Finished web page.
Changes made:
1. Removed the recentlyFinishedFlows cache.
2. Added new SQL query to fetch recentlyFinishedFlows which end within 10 mins and status is finished (SUCCEEDED/KILLED/FAILED). Execution_flows DB table is already indexed by end_time, so the query performance is good.
3. Added unit test. Also tested on solo server and testing cluster.

### CODE REVIEW

File: azkaban-common/src/main/java/azkaban/executor/ExecutorManager.java [hunk 1]
```diff
@@ -107,6 +104,8 @@
   // 12 weeks
   private static final long DEFAULT_EXECUTION_LOGS_RETENTION_MS = 3 * 4 * 7
       * 24 * 60 * 60 * 1000L;
+  // 10 mins recently finished threshold.
+  private long recentlyFinishedLifetimeMs = 600000;
```
Comment (chengren311) [inline azkaban-common/src/main/java/azkaban/executor/ExecutorManager.java]: final/static and upper case

Comment (chengren311) [inline azkaban-common/src/main/java/azkaban/executor/ExecutorManager.java]: maybe consider using duration ?

File: azkaban-common/src/main/java/azkaban/executor/JdbcExecutorLoader.java [hunk 2]
```diff
@@ -1364,6 +1382,52 @@ public String handle(ResultSet rs) throws SQLException {
...
+      do {
+        int id = rs.getInt(1);
+        int encodingType = rs.getInt(2);
+        byte[] data = rs.getBytes(3);
+
+        if (data != null) {
+          EncodingType encType = EncodingType.fromInteger(encodingType);
+          Object flowObj;
+          try {
+            if (encType == EncodingType.GZIP) {
```
Comment (chengren311) [inline azkaban-common/src/main/java/azkaban/executor/JdbcExecutorLoader.java]: this piece of code is duplicate all over the place, can we make it a util? e.x it's also used in https://github.com/azkaban/azkaban/pull/1079

File: azkaban-common/src/test/java/azkaban/executor/JdbcExecutorLoaderTest.java [hunk 3]
```diff
@@ -59,6 +59,7 @@
   private static final String user = "azkaban";
   private static final String password = "azkaban";
   private static final int numConnections = 10;
+  private static final long recentlyFinishedLifetimeMs = 5000;
```
Comment (chengren311) [inline azkaban-common/src/test/java/azkaban/executor/JdbcExecutorLoaderTest.java]: upper case. also using duration might be helpful

File: azkaban-common/src/test/java/azkaban/executor/MockExecutorLoader.java [hunk 4]
```diff
@@ -388,4 +388,10 @@ public Executor fetchExecutorByExecutionId(int execId) throws ExecutorManagerExc
   public void unassignExecutor(int executionId) throws ExecutorManagerException {
     executionExecutorMapping.remove(executionId);
   }
+
+  @Override
+  public List<ExecutableFlow> fetchRecentlyFinishedFlows(long lifeTimeMs)
+      throws ExecutorManagerException {
+    return null;
```
Comment (chengren311) [inline azkaban-common/src/test/java/azkaban/executor/MockExecutorLoader.java]: i think we should return an empty list instead of null here.

File: azkaban-common/src/main/java/azkaban/executor/ExecutorManager.java (hunk 1 above)
Comment (suvodeep-pyne) [inline azkaban-common/src/main/java/azkaban/executor/ExecutorManager.java]: Thanks @chengren311 ! 
@jamiesjc please take a look at the `Duration` class for storing time constants.

Comment (jamiesjc) [inline azkaban-common/src/main/java/azkaban/executor/ExecutorManager.java]: Fixed. Thanks!

File: azkaban-common/src/test/java/azkaban/executor/JdbcExecutorLoaderTest.java (hunk 3 above)
Comment (suvodeep-pyne) [inline azkaban-common/src/test/java/azkaban/executor/JdbcExecutorLoaderTest.java]: nit:
Will the test take 5s to finish? Will it be possible to reduce the time? 
We can mock this out if required.

File: azkaban-common/src/test/java/azkaban/executor/MockExecutorLoader.java (hunk 4 above)
Comment (suvodeep-pyne) [inline azkaban-common/src/test/java/azkaban/executor/MockExecutorLoader.java]: I think we should probably delete this class :@

File: azkaban-common/src/test/java/azkaban/executor/JdbcExecutorLoaderTest.java (hunk 3 above)
Comment (jamiesjc) [inline azkaban-common/src/test/java/azkaban/executor/JdbcExecutorLoaderTest.java]: Yes, tried reducing to 1s and it also works.

File: azkaban-common/src/test/java/azkaban/executor/MockExecutorLoader.java (hunk 4 above)
Comment (jamiesjc) [inline azkaban-common/src/test/java/azkaban/executor/MockExecutorLoader.java]: I agree. It's better to use Mockito to mock the executor loader. Currently this class is used in many test cases, but we can gradually retire it in the future.

File: azkaban-common/src/main/java/azkaban/executor/JdbcExecutorLoader.java (hunk 2 above)
Comment (jamiesjc) [inline azkaban-common/src/main/java/azkaban/executor/JdbcExecutorLoader.java]: Yes, I also noticed that. Plan to do it in a separate pull request.

File: azkaban-common/src/main/java/azkaban/executor/ExecutorManager.java [hunk 5]
```diff
@@ -107,6 +105,8 @@
   // 12 weeks
   private static final long DEFAULT_EXECUTION_LOGS_RETENTION_MS = 3 * 4 * 7
       * 24 * 60 * 60 * 1000L;
+  // 10 mins recently finished threshold.
+  private static final Duration RECENTLY_FINISHED_LIFETIME = Duration.ofSeconds(600);
```
Comment (chengren311) [inline azkaban-common/src/main/java/azkaban/executor/ExecutorManager.java]: why not ofMins(10) ?

File: azkaban-common/src/main/java/azkaban/executor/JdbcExecutorLoader.java [hunk 6]
```diff
@@ -1364,6 +1382,52 @@ public String handle(ResultSet rs) throws SQLException {
     }
   }
 
+  private static class FetchRecentlyFinishedFlows implements
```
Comment (kunkun-tang) [inline azkaban-common/src/main/java/azkaban/executor/JdbcExecutorLoader.java]: Would FetchRecentlyFinishedFlowsHandler be a better name? though I know the handler classes in this file don't have this convention. What do you think?

File: azkaban-common/src/test/java/azkaban/executor/JdbcExecutorLoaderTest.java [hunk 7]
```diff
@@ -873,6 +875,30 @@ public void testFetchActiveFlowByExecId() throws Exception {
     Assert.assertEquals(flow1.getVersion(), execFlow1.getVersion());
   }
 
+  @Test
```
Comment (kunkun-tang) [inline azkaban-common/src/test/java/azkaban/executor/JdbcExecutorLoaderTest.java]: Good test strategy!

File: azkaban-common/src/test/java/azkaban/executor/MockExecutorLoader.java [hunk 8]
```diff
@@ -388,4 +388,10 @@ public Executor fetchExecutorByExecutionId(int execId) throws ExecutorManagerExc
   public void unassignExecutor(int executionId) throws ExecutorManagerException {
     executionExecutorMapping.remove(executionId);
   }
+
+  @Override
+  public List<ExecutableFlow> fetchRecentlyFinishedFlows(long lifeTimeMs)
+      throws ExecutorManagerException {
+    return new ArrayList<>();
```
Comment (kunkun-tang) [inline azkaban-common/src/test/java/azkaban/executor/MockExecutorLoader.java]: Is Java smart enough to know the generic type here? Never knew this.

File: azkaban-common/src/main/java/azkaban/executor/JdbcExecutorLoader.java [hunk 9]
```diff
@@ -1364,6 +1382,52 @@ public String handle(ResultSet rs) throws SQLException {
...
+    ResultSetHandler<List<ExecutableFlow>> {
+    // Execution_flows table is already indexed by end_time
```
Comment (kunkun-tang) [inline azkaban-common/src/main/java/azkaban/executor/JdbcExecutorLoader.java]: A bit confused what this comment want to tell readers? Should it be "Execution_flows table should index end_time"

File: azkaban-common/src/test/java/azkaban/executor/JdbcExecutorLoaderTest.java [hunk 10]
```diff
@@ -873,6 +874,29 @@ public void testFetchActiveFlowByExecId() throws Exception {
     Assert.assertEquals(flow1.getVersion(), execFlow1.getVersion());
```
Comment (HappyRay) [inline azkaban-common/src/test/java/azkaban/executor/JdbcExecutorLoaderTest.java]: See http://stackoverflow.com/questions/1371940/unit-testing-is-it-a-good-practice-to-have-assertions-in-setup-methods

File: azkaban-common/src/test/java/azkaban/executor/JdbcExecutorLoaderTest.java [hunk 11]
```diff
@@ -873,6 +874,29 @@ public void testFetchActiveFlowByExecId() throws Exception {
...
+    flow1.setStatus(Status.SUCCEEDED);
+    flow1.setEndTime(System.currentTimeMillis());
+    loader.updateExecutableFlow(flow1);
+    List<ExecutableFlow> flows = loader.fetchRecentlyFinishedFlows(recentlyFinishedLifetimeMs);
+    Assert.assertEquals(1, flows.size());
+    Assert.assertEquals(flow1.getExecutionId(), flows.get(0).getExecutionId());
+    Assert.assertEquals(flow1.getProjectName(), flows.get(0).getProjectName());
+    Assert.assertEquals(flow1.getFlowId(), flows.get(0).getFlowId());
+    Assert.assertEquals(flow1.getVersion(), flows.get(0).getVersion());
+    Thread.currentThread().sleep(recentlyFinishedLifetimeMs);
```
Comment (HappyRay) [inline azkaban-common/src/test/java/azkaban/executor/JdbcExecutorLoaderTest.java]: Why does the thread need to sleep here? 
It's important to keep a unit test run as fast as possible. in a few ms is the norm.

File: azkaban-common/src/test/java/azkaban/executor/JdbcExecutorLoaderTest.java [hunk 12]
```diff
@@ -873,6 +874,29 @@ public void testFetchActiveFlowByExecId() throws Exception {
     Assert.assertEquals(flow1.getVersion(), execFlow1.getVersion());
   }
 
+  @Test
+  public void testFetchRecentlyFinishedFlows() throws Exception {
+    if (!isTestSetup()) {
```
Comment (HappyRay) [inline azkaban-common/src/test/java/azkaban/executor/JdbcExecutorLoaderTest.java]: My understanding is that isTestSetup will return false in Travis effectively disabling this test. Am I right?

File: azkaban-common/src/test/java/azkaban/executor/JdbcExecutorLoaderTest.java [hunk 13]
```diff
@@ -873,6 +874,29 @@ public void testFetchActiveFlowByExecId() throws Exception {
...
+    flow1.setEndTime(System.currentTimeMillis());
+    loader.updateExecutableFlow(flow1);
+    List<ExecutableFlow> flows = loader.fetchRecentlyFinishedFlows(recentlyFinishedLifetimeMs);
+    Assert.assertEquals(1, flows.size());
+    Assert.assertEquals(flow1.getExecutionId(), flows.get(0).getExecutionId());
+    Assert.assertEquals(flow1.getProjectName(), flows.get(0).getProjectName());
+    Assert.assertEquals(flow1.getFlowId(), flows.get(0).getFlowId());
+    Assert.assertEquals(flow1.getVersion(), flows.get(0).getVersion());
+    Thread.currentThread().sleep(recentlyFinishedLifetimeMs);
+    flows = loader.fetchRecentlyFinishedFlows(recentlyFinishedLifetimeMs);
```
Comment (HappyRay) [inline azkaban-common/src/test/java/azkaban/executor/JdbcExecutorLoaderTest.java]: Have you considered setting up an in-memory DB state and just test the fetch method? 
@kunkun-tang has introduced some tests using in-memory DB.

File: azkaban-common/src/test/java/azkaban/executor/JdbcExecutorLoaderTest.java (hunk 10 above)
Comment (jamiesjc) [inline azkaban-common/src/test/java/azkaban/executor/JdbcExecutorLoaderTest.java]: This assert is not inside setupDB(). It's in a separate test case.

File: azkaban-common/src/test/java/azkaban/executor/JdbcExecutorLoaderTest.java (hunk 12 above)
Comment (jamiesjc) [inline azkaban-common/src/test/java/azkaban/executor/JdbcExecutorLoaderTest.java]: Yes, in order to run all these tests in JdbcExecutorLoaderTest class locally, we need to modify the database, user and password to match our local mysql setup. In Travis, all the tests will be disabled because DB is not set up. Once we change it to in-memory database later, this JdbcExecutorLoaderTest class will be refactored and the tests won't have this setup limitation.

File: azkaban-common/src/test/java/azkaban/executor/MockExecutorLoader.java (hunk 8 above)
Comment (jamiesjc) [inline azkaban-common/src/test/java/azkaban/executor/MockExecutorLoader.java]: Yes, from the return type it is able to figure out. If you explicitly put the type ArrayList<ExecutableFlow>, intellij will suggest you to remove it.

File: azkaban-common/src/main/java/azkaban/executor/JdbcExecutorLoader.java (hunk 9 above)
Comment (jamiesjc) [inline azkaban-common/src/main/java/azkaban/executor/JdbcExecutorLoader.java]: I added this comment in case readers have doubt about the query performance.

File: azkaban-common/src/test/java/azkaban/executor/JdbcExecutorLoaderTest.java (hunk 7 above)
Comment (jamiesjc) [inline azkaban-common/src/test/java/azkaban/executor/JdbcExecutorLoaderTest.java]: Thanks! : )

File: azkaban-common/src/test/java/azkaban/executor/JdbcExecutorLoaderTest.java (hunk 10 above)
Comment (HappyRay) [inline azkaban-common/src/test/java/azkaban/executor/JdbcExecutorLoaderTest.java]: Got it. I was confused by the diff display.

File: azkaban-common/src/main/java/azkaban/executor/ExecutorManager.java (hunk 5 above)
Comment (HappyRay) [inline azkaban-common/src/main/java/azkaban/executor/ExecutorManager.java]: +1

File: azkaban-common/src/test/java/azkaban/executor/JdbcExecutorLoaderTest.java [hunk 14]
```diff
@@ -873,6 +875,30 @@ public void testFetchActiveFlowByExecId() throws Exception {
...
+    flow1.setEndTime(System.currentTimeMillis());
+    loader.updateExecutableFlow(flow1);
+    List<ExecutableFlow> flows = loader.fetchRecentlyFinishedFlows(
+        RECENTLY_FINISHED_LIFETIME.toMillis());
+    Assert.assertEquals(1, flows.size());
+    Assert.assertEquals(flow1.getExecutionId(), flows.get(0).getExecutionId());
+    Assert.assertEquals(flow1.getProjectName(), flows.get(0).getProjectName());
+    Assert.assertEquals(flow1.getFlowId(), flows.get(0).getFlowId());
+    Assert.assertEquals(flow1.getVersion(), flows.get(0).getVersion());
+    Thread.currentThread().sleep(RECENTLY_FINISHED_LIFETIME.toMillis());
```
Comment (HappyRay) [inline azkaban-common/src/test/java/azkaban/executor/JdbcExecutorLoaderTest.java]: We would like unit tests to run as fast as possible.
It seems that in this test you can control the flow end time. If you set an end time e.g. 20 minutes in the past, the test doesn't need to sleep, does it?

Also I don't see a positive test case, i.e. there is a flow that finishes within 10 minutes.

File: azkaban-common/src/main/java/azkaban/executor/ExecutorManager.java (hunk 5 above)
Comment (jamiesjc) [inline azkaban-common/src/main/java/azkaban/executor/ExecutorManager.java]: Changed. Thanks!

File: azkaban-common/src/test/java/azkaban/executor/JdbcExecutorLoaderTest.java (hunk 14 above)
Comment (jamiesjc) [inline azkaban-common/src/test/java/azkaban/executor/JdbcExecutorLoaderTest.java]: Yes, I've tried using DateTimeUtils to set end time in the past, and it works. Thanks!
In line 890 within the same test, it is the positive scenario where it successfully fetches the recently finished flow. I will split the two tests for clarity.

File: azkaban-common/src/main/java/azkaban/executor/ExecutorLoader.java [hunk 15]
```diff
@@ -32,6 +32,9 @@ void uploadExecutableFlow(ExecutableFlow flow)
   ExecutableFlow fetchExecutableFlow(int execId)
       throws ExecutorManagerException;
 
+  List<ExecutableFlow> fetchRecentlyFinishedFlows(long lifeTimeMs)
```
Comment (HappyRay) [inline azkaban-common/src/main/java/azkaban/executor/ExecutorLoader.java]: It's not obvious what the lifeTimeMs means. Add a javadoc?

Does the name like "maxAge" more closely reflect the intent?

How about using the Duration instead here and convert to long when needed? At this level of abstraction, users of this API shouldn't share how this value is represented in the DB.

File: azkaban-common/src/main/java/azkaban/executor/ExecutorManager.java [hunk 16]
```diff
@@ -641,7 +641,14 @@ private void getRunningFlowsIdsHelper(List<Integer> allIds,
 
   @Override
   public List<ExecutableFlow> getRecentlyFinishedFlows() {
-    return new ArrayList<ExecutableFlow>(recentlyFinished.values());
+    List<ExecutableFlow> flows = new ArrayList<>();
+    try {
+      flows = executorLoader.fetchRecentlyFinishedFlows(
+          RECENTLY_FINISHED_LIFETIME.toMillis());
+    } catch(ExecutorManagerException e) {
+      logger.error(e);
```
Comment (HappyRay) [inline azkaban-common/src/main/java/azkaban/executor/ExecutorManager.java]: This will call
void error(Object message) {
which will not log the exception details such as call stack.
Always use this method to log an exception. 
void error(Object message, Throwable t) {

Comment (HappyRay) [inline azkaban-common/src/main/java/azkaban/executor/ExecutorManager.java]: Should the server return a server error (5xx) instead?

File: azkaban-common/src/main/java/azkaban/executor/JdbcExecutorLoader.java [hunk 17]
```diff
@@ -1364,6 +1382,52 @@ public String handle(ResultSet rs) throws SQLException {
...
+    private static String FETCH_RECENTLY_FINISHED_FLOW =
```
Comment (HappyRay) [inline azkaban-common/src/main/java/azkaban/executor/JdbcExecutorLoader.java]: What do you think about limiting the returned result to the most recent N flows instead regardless of how long ago they finish?

File: azkaban-common/src/main/java/azkaban/executor/ExecutorLoader.java (hunk 15 above)
Comment (jamiesjc) [inline azkaban-common/src/main/java/azkaban/executor/ExecutorLoader.java]: Changed. Thanks!

File: azkaban-common/src/main/java/azkaban/executor/ExecutorManager.java (hunk 16 above)
Comment (jamiesjc) [inline azkaban-common/src/main/java/azkaban/executor/ExecutorManager.java]: Changed to error(Object message, Throwable t). I think it's ok to just keep the existing behavior to return an empty list if it fails to fetch recently finished flows. What is the intention of returning a server error?

File: azkaban-common/src/main/java/azkaban/executor/JdbcExecutorLoader.java (hunk 17 above)
Comment (jamiesjc) [inline azkaban-common/src/main/java/azkaban/executor/JdbcExecutorLoader.java]: That's also an option. But I think it's a little hard to decide the exact number limit for all clusters. On some clusters we might want to display more? Also I guess users have already got used to look at the flows finished in last 10 mins?

File: azkaban-common/src/main/java/azkaban/executor/ExecutorManager.java (hunk 16 above)
Comment (HappyRay) [inline azkaban-common/src/main/java/azkaban/executor/ExecutorManager.java]: Returning a server error will help with the monitoring system for example.

File: azkaban-common/src/main/java/azkaban/executor/ExecutorLoader.java (hunk 15 above)
Comment (HappyRay) [inline azkaban-common/src/main/java/azkaban/executor/ExecutorLoader.java]: I don't see the change yet. Did you submit it?

File: azkaban-common/src/main/java/azkaban/executor/ExecutorManager.java [hunk 18]
```diff
@@ -107,6 +105,8 @@
   // 12 weeks
   private static final long DEFAULT_EXECUTION_LOGS_RETENTION_MS = 3 * 4 * 7
       * 24 * 60 * 60 * 1000L;
+  // 10 mins recently finished threshold.
```
Comment (HappyRay) [inline azkaban-common/src/main/java/azkaban/executor/ExecutorManager.java]: Is this comment still necessary? The code seems to be clear enough.

File: azkaban-common/src/main/java/azkaban/executor/ExecutorManager.java [hunk 19]
```diff
@@ -641,7 +641,14 @@ private void getRunningFlowsIdsHelper(List<Integer> allIds,
 
   @Override
   public List<ExecutableFlow> getRecentlyFinishedFlows() {
-    return new ArrayList<ExecutableFlow>(recentlyFinished.values());
+    List<ExecutableFlow> flows = new ArrayList<>();
+    try {
+      flows = executorLoader.fetchRecentlyFinishedFlows(
+          RECENTLY_FINISHED_LIFETIME);
+    } catch(ExecutorManagerException e) {
+      logger.error("Failed to fetch recently finished flows.", e);
```
Comment (HappyRay) [inline azkaban-common/src/main/java/azkaban/executor/ExecutorManager.java]: I would suggest adding a todo here to fix the error handling.

File: azkaban-common/src/main/java/azkaban/executor/JdbcExecutorLoader.java [hunk 20]
```diff
@@ -1364,6 +1386,52 @@ public String handle(ResultSet rs) throws SQLException {
     }
   }
 
+  private static class FetchRecentlyFinishedFlows implements
```
Comment (HappyRay) [inline azkaban-common/src/main/java/azkaban/executor/JdbcExecutorLoader.java]: Another strategy is to break up a larger pull request into multiple smaller ones.  Create an issue to tie them together. This can speed up the review process. In case there is a need to rollback, you have an option to roll back a portion of the changes.

e.g. adding a new method like this can go in first independently. Since it is not used yet, it is safe.

File: azkaban-common/src/test/java/azkaban/executor/JdbcExecutorLoaderTest.java [hunk 21]
```diff
@@ -873,6 +877,50 @@ public void testFetchActiveFlowByExecId() throws Exception {
...
+    }
+
+    ExecutorLoader loader = createLoader();
+    ExecutableFlow flow1 = TestUtils.createExecutableFlow("exectest1", "exec1");
+    loader.uploadExecutableFlow(flow1);
+    flow1.setStatus(Status.SUCCEEDED);
+    flow1.setEndTime(DateTimeUtils.currentTimeMillis());
+    loader.updateExecutableFlow(flow1);
+    //Mock flow finished time to be 2 min ago.
+    DateTimeUtils.setCurrentMillisOffset(-FLOW_FINISHED_TIME.toMillis());
```
Comment (HappyRay) [inline azkaban-common/src/test/java/azkaban/executor/JdbcExecutorLoaderTest.java]: Is it necessary to modify the current time? Isn't create a new time enough?

We should use the java8 java.time api instead of jodatime in the new code. 
https://stackoverflow.com/questions/29750221/is-joda-time-deprecated-with-java-8-date-and-time-api-java-time
